File: pairing-service/client-examples/remote_pairing_client.dart
import 'dart:async';
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:web_socket_channel/web_socket_channel.dart';
import 'package:web_socket_channel/status.dart' as status;

/// Credentials model for Xtream Codes
class XtreamCredentials {
  final String serverUrl;
  final String username;
  final String password;

  const XtreamCredentials({
    required this.serverUrl,
    required this.username,
    required this.password,
  });

  factory XtreamCredentials.fromJson(Map<String, dynamic> json) {
    return XtreamCredentials(
      serverUrl: json['serverUrl'] as String? ?? '',
      username: json['username'] as String? ?? '',
      password: json['password'] as String? ?? '',
    );
  }
}

/// Provisioning payload received from the Admin Web Portal
class ProvisionPayload {
  final String playlistType; // "m3u" or "xtream"
  final String? playlistUrl;
  final String? epgUrl;
  final XtreamCredentials? xtreamData;

  const ProvisionPayload({
    required this.playlistType,
    this.playlistUrl,
    this.epgUrl,
    this.xtreamData,
  });

  factory ProvisionPayload.fromJson(Map<String, dynamic> json) {
    return ProvisionPayload(
      playlistType: json['playlistType'] as String? ?? 'm3u',
      playlistUrl: json['playlistUrl'] as String?,
      epgUrl: json['epgUrl'] as String?,
      xtreamData: json['xtreamData'] != null
          ? XtreamCredentials.fromJson(json['xtreamData'] as Map<String, dynamic>)
          : null,
    );
  }
}

/// Active session metadata
class PairingSession {
  final String code;
  final int expiresInSeconds;
  final String qrUrl;
  final String webSocketUrl;

  const PairingSession({
    required this.code,
    required this.expiresInSeconds,
    required this.qrUrl,
    required this.webSocketUrl,
  });
}

/// Production-ready Remote Pairing Service for Flutter IPTV apps
class RemotePairingService {
  final http.Client _httpClient;
  WebSocketChannel? _channel;
  StreamSubscription? _subscription;
  bool _isDisposed = false;

  RemotePairingService({http.Client? httpClient})
      : _httpClient = httpClient ?? http.Client();

  /// Initiates the pairing session and connects the real-time WebSocket.
  ///
  /// [serverBaseUrl] should be e.g. "https://pair.opentv.app" or "http://192.168.1.50:3000".
  /// [onSessionReady] is triggered when the 6-character code and QR URL are received.
  /// [onProvisioned] is triggered when the admin submits credentials.
  /// [onError] reports network or validation issues.
  Future<void> startPairing({
    required String serverBaseUrl,
    required void Function(PairingSession session) onSessionReady,
    required void Function(ProvisionPayload payload) onProvisioned,
    required void Function(String error) onError,
    void Function()? onConnected,
  }) async {
    _isDisposed = false;
    final cleanBaseUrl = serverBaseUrl.replaceAll(RegExp(r'/+$'), '');

    // Step 1: Hit POST /api/pair/init
    try {
      final response = await _httpClient.post(
        Uri.parse('$cleanBaseUrl/api/pair/init'),
        headers: {
          'Content-Type': 'application/json',
          'Accept': 'application/json',
        },
        body: jsonEncode({}),
      );

      if (_isDisposed) return;

      if (response.statusCode != 200) {
        onError('Failed to initialize session: HTTP ${response.statusCode}');
        return;
      }

      final data = jsonDecode(response.body) as Map<String, dynamic>;
      final code = data['code'] as String;
      final expiresIn = data['expiresIn'] as int? ?? 600;

      final qrUrl = '$cleanBaseUrl/?code=$code';

      // Format WebSocket URL (ws:// or wss://)
      final isSecure = cleanBaseUrl.toLowerCase().startsWith('https://');
      final host = cleanBaseUrl.replaceFirst(RegExp(r'^https?://'), '');
      final wsUrl = '${isSecure ? 'wss' : 'ws'}://$host/?code=$code';

      final session = PairingSession(
        code: code,
        expiresInSeconds: expiresIn,
        qrUrl: qrUrl,
        webSocketUrl: wsUrl,
      );

      onSessionReady(session);

      // Step 2: Open WebSocket connection
      _connectWebSocket(
        wsUrl: wsUrl,
        onConnected: onConnected,
        onProvisioned: onProvisioned,
        onError: onError,
      );
    } catch (e) {
      if (!_isDisposed) {
        onError('Network error during session initialization: $e');
      }
    }
  }

  void _connectWebSocket({
    required String wsUrl,
    void Function()? onConnected,
    required void Function(ProvisionPayload payload) onProvisioned,
    required void Function(String error) onError,
  }) {
    try {
      final channel = WebSocketChannel.connect(Uri.parse(wsUrl));
      _channel = channel;

      _subscription = channel.stream.listen(
        (dynamic rawMessage) {
          if (_isDisposed) return;
          try {
            final json = jsonDecode(rawMessage.toString()) as Map<String, dynamic>;
            final type = json['type'] as String?;

            if (type == 'connected') {
              onConnected?.call();
            } else if (type == 'provision') {
              final payload = ProvisionPayload.fromJson(json);
              onProvisioned(payload);
              // Cleanly close after receiving provision data
              dispose();
            }
          } catch (e) {
            onError('Error parsing incoming configuration: $e');
          }
        },
        onError: (error) {
          if (!_isDisposed) {
            onError('WebSocket communication error: $error');
          }
        },
        onDone: () {
          // Socket closed by server or network
        },
        cancelOnError: true,
      );
    } catch (e) {
      if (!_isDisposed) {
        onError('Could not connect to WebSocket: $e');
      }
    }
  }

  /// Cancel active subscriptions and close open sockets
  void dispose() {
    _isDisposed = true;
    _subscription?.cancel();
    _subscription = null;
    _channel?.sink.close(status.goingAway);
    _channel = null;
  }
}

File: pairing-service/client-examples/RemotePairingClient.kt
package app.opentv.pairing

import android.os.Handler
import android.os.Looper
import android.util.Log
import okhttp3.Call
import okhttp3.Callback
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import org.json.JSONObject
import java.io.IOException
import java.util.concurrent.TimeUnit

/**
 * Production-ready client for OpenTV Remote Pairing & Provisioning Service.
 *
 * Workflow:
 * 1. Calls `POST /api/pair/init` to obtain an ephemeral 6-character code.
 * 2. Formats a QR code URL: `https://<YOUR_DOMAIN>/?code=<CODE>`.
 * 3. Opens a WebSocket to `wss://<YOUR_DOMAIN>/?code=<CODE>`.
 * 4. Listens for real-time provisioning payload (`m3u` or `xtream`).
 * 5. Passes extracted credentials to callback and closes the session.
 */
class RemotePairingClient(
    private val client: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(0, TimeUnit.MILLISECONDS) // Indefinite for WebSocket
        .pingInterval(25, TimeUnit.SECONDS)
        .build()
) {

    data class PairingSession(
        val code: String,
        val expiresInSeconds: Int,
        val qrUrl: String,
        val webSocketUrl: String
    )

    data class XtreamCredentials(
        val serverUrl: String,
        val username: String,
        val password: String
    )

    data class ProvisionPayload(
        val playlistType: String, // "m3u" or "xtream"
        val playlistUrl: String?,
        val epgUrl: String?,
        val xtreamData: XtreamCredentials?
    )

    interface PairingCallback {
        fun onSessionReady(session: PairingSession)
        fun onConnected()
        fun onProvisioned(payload: ProvisionPayload)
        fun onError(errorMessage: String)
        fun onClosed()
    }

    private var currentWebSocket: WebSocket? = null
    private val mainHandler = Handler(Looper.getMainLooper())
    @Volatile private var isCancelled = false

    /**
     * Start the remote pairing process.
     *
     * @param serverBaseUrl Base URL of your self-hosted service, e.g. "https://pair.opentv.app" or "http://192.168.1.50:3000"
     * @param callback Status and result callbacks invoked on the Android Main Thread.
     */
    fun startPairing(serverBaseUrl: String, callback: PairingCallback) {
        isCancelled = false
        val cleanBaseUrl = serverBaseUrl.trimEnd('/')

        val initUrl = "$cleanBaseUrl/api/pair/init"
        val request = Request.Builder()
            .url(initUrl)
            .post("{}".toRequestBody("application/json; charset=utf-8".toMediaType()))
            .header("Accept", "application/json")
            .build()

        client.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                if (isCancelled) return
                postError(callback, "Failed to connect to pairing service: ${e.localizedMessage}")
            }

            override fun onResponse(call: Call, response: Response) {
                if (isCancelled) return
                response.use { res ->
                    if (!res.isSuccessful) {
                        postError(callback, "Server returned error: HTTP ${res.code}")
                        return
                    }

                    val bodyStr = res.body?.string().orEmpty()
                    try {
                        val json = JSONObject(bodyStr)
                        val code = json.getString("code")
                        val expiresIn = json.optInt("expiresIn", 600)

                        val qrUrl = "$cleanBaseUrl/?code=$code"

                        // Convert http(s) to ws(s)
                        val wsProtocol = if (cleanBaseUrl.startsWith("https://", ignoreCase = true)) "wss" else "ws"
                        val hostAndPort = cleanBaseUrl.substringAfter("://")
                        val wsUrl = "$wsProtocol://$hostAndPort/?code=$code"

                        val session = PairingSession(
                            code = code,
                            expiresInSeconds = expiresIn,
                            qrUrl = qrUrl,
                            webSocketUrl = wsUrl
                        )

                        mainHandler.post { callback.onSessionReady(session) }

                        // Step 3: Open the WebSocket connection
                        connectWebSocket(wsUrl, callback)

                    } catch (e: Exception) {
                        postError(callback, "Failed to parse session response: ${e.localizedMessage}")
                    }
                }
            }
        })
    }

    private fun connectWebSocket(wsUrl: String, callback: PairingCallback) {
        if (isCancelled) return

        val wsRequest = Request.Builder()
            .url(wsUrl)
            .build()

        currentWebSocket = client.newWebSocket(wsRequest, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                Log.d(TAG, "WebSocket connection opened to pairing server.")
                mainHandler.post { callback.onConnected() }
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                Log.d(TAG, "WebSocket message received: $text")
                try {
                    val json = JSONObject(text)
                    val type = json.optString("type")

                    if (type == "provision") {
                        val playlistType = json.optString("playlistType", "m3u")
                        val playlistUrl = json.optString("playlistUrl").takeIf { it.isNotEmpty() && it != "null" }
                        val epgUrl = json.optString("epgUrl").takeIf { it.isNotEmpty() && it != "null" }

                        var xtreamCreds: XtreamCredentials? = null
                        if (json.has("xtreamData") && !json.isNull("xtreamData")) {
                            val xObj = json.getJSONObject("xtreamData")
                            xtreamCreds = XtreamCredentials(
                                serverUrl = xObj.getString("serverUrl"),
                                username = xObj.getString("username"),
                                password = xObj.getString("password")
                            )
                        }

                        val payload = ProvisionPayload(
                            playlistType = playlistType,
                            playlistUrl = playlistUrl,
                            epgUrl = epgUrl,
                            xtreamData = xtreamCreds
                        )

                        mainHandler.post {
                            callback.onProvisioned(payload)
                        }

                        // Close socket after successful delivery
                        webSocket.close(1000, "Provisioning data received by client")
                    }
                } catch (e: Exception) {
                    Log.e(TAG, "Error handling incoming payload", e)
                }
            }

            override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
                webSocket.close(1000, null)
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                Log.d(TAG, "WebSocket closed ($code: $reason)")
                mainHandler.post { callback.onClosed() }
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                if (isCancelled) return
                Log.w(TAG, "WebSocket failure", t)
                postError(callback, "Connection dropped: ${t.localizedMessage}")
            }
        })
    }

    /**
     * Stop and cancel any active pairing session.
     */
    fun cancel() {
        isCancelled = true
        try {
            currentWebSocket?.close(1000, "User cancelled pairing")
        } catch (_: Exception) {}
        currentWebSocket = null
    }

    private fun postError(callback: PairingCallback, msg: String) {
        if (!isCancelled) {
            mainHandler.post { callback.onError(msg) }
        }
    }

    companion object {
        private const val TAG = "RemotePairingClient"
    }
}
